### PR TITLE
README edits

### PR DESCRIPTION
**Pull request by zhekson1**
WIP README edits

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
+
+-----
+
 ## Motivation
-Any DEX that operates on a Proof-of-Stake (PoS) blockchain **MUST** include delegation control as a foundational feature for the protocol. It is existentially important for the long-term sustainability of the underlying blockchain. Yet, all Cardano DEXs either cannot give users full delegation control by design or delegation control is just an afterthought.
+Many DEXes on Cardano are currently implemented in ways that lock users' assets into a tightly fixed, and/or centrally maintained, set(s) of script addresses. Such design patterns are reminiscent of the EVM's accounts-based programming paradigm, and inherit many of the same downsides, namely scalability bottlenecks and asset centralization. DEXes that hope to achieve massive scale on the CSL must adhere to a radically different approach that takes full advantage of the concurrency and parallelism offered by the eUTxO model. `cardano-swaps` is a first attempt at such an approach.
```
[the remark this answers is not in the document]
Comment (fallen-icarus) [reply README.md:45]: I don't know if it is accurate to say that other DEX designs are less scalable. Using batchers, they are as scalable as the underlying blockchain. This is no different than the cardano-swaps design; it is just that cardano-swaps doesn't need batchers to achieve this. I think you should instead stress that it is scalable without needing batchers instead of saying it is "actually scalable".

[the remark this answers is not in the document]
Comment (fallen-icarus) [reply README.md:45]: What about this (changes bolded):

Many DEXes on Cardano are currently implemented in ways that lock users' assets into a tightly fixed, and/or centrally maintained, set(s) of script addresses. Such design patterns are reminiscent of the EVM's accounts-based programming paradigm, and inherit many of the same downsides, namely scalability bottlenecks **that require batchers to solve** and asset **/stake** centralization. DEXes that hope to achieve massive scale on the CSL **in a truly decentralized way** must adhere to a radically different approach that takes full advantage of the concurrency and parallelism offered by the eUTxO model. `cardano-swaps` is a first attempt at such an approach.

File: README.md [hunk 2]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
+Many DEXes on Cardano are currently implemented in ways that lock users' assets into a tightly fixed, and/or centrally maintained, set(s) of script addresses. Such design patterns are reminiscent of the EVM's accounts-based programming paradigm, and inherit many of the same downsides, namely scalability bottlenecks and asset centralization. DEXes that hope to achieve massive scale on the CSL must adhere to a radically different approach that takes full advantage of the concurrency and parallelism offered by the eUTxO model. `cardano-swaps` is a first attempt at such an approach. 
+
+-----
+
+## Preliminary Discussion
+To appreciate the necessity of new DEX standards, it first important to understand the deficiencies of the current status-quo:
+
+### Current DEX Deficiencies  
+One consequence of centralized script addresses is the necessity for liquidity pools and LP providers as discrete entities. LPs are a common feature of many DEXes, coming with undesirable properties and corresponding sets of workaround "solutions". However, these workarounds have issues themselves, as explored here:
```
Comment (fallen-icarus) [inline README.md:53]: You should probably introduce LP after liquidity pools with (LP)

File: README.md [hunk 3]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
 - [Upgradability](#upgradability)
-- [Frontend Agnostic](#frontend-agnostic)
+- [Frontend Agnosticism](<#frontend agnosticism>)
 - [Conclusion](#conclusion)
 - [FAQ](#faq)
   - [Why Are Reference Scripts Required?](#if-all-users-have-direct-access-to-the-spending-script-for-that-trading-pair-why-are-reference-scripts-required)
 
 ---
+## Abstract
+`cardano-swaps` is a proof-of-concept implementation of massively scalable and concurrent DEX protocol for the Cardano Settlement Layer (CSL). It solves many of the pitfalls of current DEX implementations by empowering users to deploy their own (and interact with each others') script addresses. By doing so, users maintain fully custody of their assets at all times, and can elect if/when to upgrade their addresses to new contract standards.
```
Comment (fallen-icarus) [inline README.md:40]: "massively" can be seen as hyperbolic. Instead, use "organically".

Comment (fallen-icarus) [reply README.md:40]: You need to mention delegation control here. How about this (changes bolded):

By doing so, users maintain **full custody and delegation control** of their assets at all times, and can elect if/when to upgrade their addresses to new contract standards.

BTW, you said "fully custody"

File: README.md [hunk 4]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
+| Impermanent Loss | Yield Farming & Concentrated Liquidity | Medium - Long term unsustainability |
+| Incomplete or No Delegation Control | <ul><li>Asset pool fractionalization</li><li>Indirect Delegation via Governance Tokens</li></ul> | <ul><li>Unfair distribution of Governance tokens </li><li>Unavoidable centralization of stake (Major issue for Ouroboros)</li></ul>
+| Scaling Bottlenecks | Batchers, Execution/Routing Engines, and/or other middlemen | Middlemen can take advantage of their position between users and the protocol. Even if MEV is mitigated, more users --> more execution demand --> centralization of middlemen |
 
-Each of these downsides have their own corresponding "solution": yield farming, governance tokens, and batchers, respectively. However, these "solutions" have downsides of their own:
- 
-  1. The long term sustainability of yield farming is questionable.
-  2. Fairly distributing governance tokens is difficult.
-  3. Batchers are effectively middle-men that can take advantage of their position between users and the protocol.
+The key takeaway here (especially from the last point) is even
```
Comment (fallen-icarus) [inline README.md:61]: Did you mean to remove this sentence?

File: README.md [hunk 5]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
 
-Each of these downsides have their own corresponding "solution": yield farming, governance tokens, and batchers, respectively. However, these "solutions" have downsides of their own:
- 
-  1. The long term sustainability of yield farming is questionable.
-  2. Fairly distributing governance tokens is difficult.
-  3. Batchers are effectively middle-men that can take advantage of their position between users and the protocol.
+The key takeaway here (especially from the last point) is even 
 
-Even more concerning is that, even if governance tokens are fairly distributed, this doesn't change the fact that a liquidity pool can only be delegated to one stake pool. Fractionalizing liquidity pools helps but it still results in a more centralized PoS blockchain than if each user was able to delegate their assets independently.
+Of course, this is not an exhaustive list, but here is the key takeaway: **even if users control their own keys, for a DEX to achieve true scale the users must also control their own *addresses*.**
```
Comment (fallen-icarus) [inline README.md:63]: > **even if users control their own keys, for a DEX to achieve true scale the users must also control their own *addresses*.**

Make this "achieve true scale **in a decentralized manner**". As stated before, batchers make the centralized design just as scalable as cardano-swaps (possibly more so with the redundant executions).

File: README.md [hunk 6]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
+
+
+if some workarounds *can* be effective, the underlying design *principles* are inadequate. Much like Bittorrent, and the CSL-CCL stack, **the best p2p protocols are ones that scale in proportion to the number of participants**. DEXes are no different. With this in mind, we can explore alternative design principles:
```
Comment (fallen-icarus) [inline README.md:66]: I'm guessing this is the other half of the sentence before? I think you can just start here with "Much like Bittorrent..."

Comment (fallen-icarus) [reply README.md:66]: Shouldn't the paragraph end with a period instead of a colon?

File: README.md [hunk 7]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
-Programmable Swaps are a more promising design than liquidity pools and were first proposed by Axo (formerly known as Maladex). However, in Axo's [whitepaper](https://www.axo.trade/whitepaper.pdf), there is no mention of giving users full delegation control of their own assets. When they were asked on Discord about it, they said it was possible but delegation control would not be included in the first version.
+Programmable Swaps are a better design choice than liquidity pools. They were first proposed by Axo (formerly known as Maladex). However, in Axo's original [whitepaper](https://www.axo.trade/whitepaper.pdf), there is no mention of giving users *full* delegation control of their own assets. 
 
-### The Cardano-Swaps DEX
-Cardano-Swaps took inspiration from Axo's programmable swaps design but added delegation control as a foundational feature. `cardano-swaps` is the name of the CLI program included to help use the DEX.
+### The cardano-swaps Protocol
+cardano-swaps took inspiration from Axo's programmable swaps design but added delegation control as a foundational feature. `cardano-swaps` is the name of the CLI program that enables use of the DEX protocol.
```
Comment (fallen-icarus) [inline README.md:72]: Should "cardano-swaps" be capitalized here? The README header has it capitalized. I originally used Cardano-Swaps as the name of the DEX and `cardano-swaps` as the name of the helper CLI.

File: README.md [hunk 8]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
 
-A cardano address is made up of both a payment credential and a staking credential. As long as the staking credential is unique to that user, that user will maintain full delegation control for that address. With this in mind, every user is given the same spending script to use for the address' payment credential and are required to use their own staking credential (either a pubkey or a staking script). To force the use of a staking credential, the DEX is designed so that it is not possible to mint a beacon token (discussed later) to an address without a staking credential.
+A cardano address is made up of both a payment credential and a staking credential. As long as the staking credential is unique to that user, full delegation control over the address is maintained. With this in mind, every user is given the same spending script to use for the address' payment credential and are required to use their own staking credential (either a pubkey or a staking script). To force the use of a staking credential, the DEX is designed such that it is not possible to mint a valid beacon token (discussed later) to an address without a staking credential.
```
Comment (fallen-icarus) [inline README.md:94]: Drop the "valid" here for "mint a *valid* beacon token" since it isn't possible to mint anything if the address doesn't have a staking credential.

File: README.md [hunk 9]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
 
 - If the staking credential is a pubkey, then the staking pubkey must sign the transaction.
 - If the staking credential is a script, then the script must be executed in the same transaction.
+- the staking credential effectively *becomes* the "owner" for all actions except for the actual swap execution, in which case the spending credential is used directly.
 
-:note: It is possible to execute a staking script even if 0 ADA is withrawn from a rewards address. The only requirement to use staking scripts like this is that the associated stake address must be registered and delegated. **You can start using the stake address like this the moment the registration+delegation transaction gets added to the chain. You do not need to wait an epoch.**
+ It is possible to execute a staking script even if 0 ADA is withdrawn from a reward address. The only requirement to use staking scripts like this is that the associated stake address must be registered and delegated. Stake addresses can be utilized this way as soon as the registration+delegation transaction is added to the chain. There is no epoch waiting-period.
```
Comment (fallen-icarus) [inline README.md:103]: This is a slight tangent which is why I prefaced it with :note: . Github converts :note: it to an image ([reference](https://stackoverflow.com/questions/58737436/how-to-create-a-good-looking-notification-or-warning-box-in-github-flavoured-mar)).

File: README.md [hunk 10]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
+#### Why not just give each user their own spending script?
+`cardano-swaps` relies on the usage of [beacon tokens](<#specification/##Beacon Tokens>) to "tag" addresses, which demarcates them as *distinct* for efficient off-chain querying/aggregation. However, **address distinction is nonsensical if each script address is composed of completely unique credentials.** In other words, for Beacon Tokens to work properly, there must be a clear distinction between addresses participating in a given dApp and all other addresses. `cardano-swaps` leverages the dual payment-staking credentials of Cardano addresses to maintain address-distinction (shared spending scripts) without sacrificing self-custody (unique staking keys/scripts). For further clarification, refer to the [Beacon Token](<#specification/##Beacon Tokens>) section below, or to the Beacon Token CIP (**link here**)
```
Comment (fallen-icarus) [inline README.md:106]: Perhaps *impossible* is more appropriate than *nonsensical*.

File: README.md [hunk 11]
```diff
@@ -30,71 +30,88 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
 
-### Delegating
-With each user having their own DEX address, delegation the address is identical to delegating a traditional payment address that uses either a staking pubkey or a staking script. You can check out [delegation section](GettingStarted.md#delegate-the-swap-address) of GettingStarted.md for an example using a staking pubkey.
+:note v1.0.0 of cardano-swaps used unique spending scripts; you can read about the limitations in the v1.0.0 commit README.
```
Comment (fallen-icarus) [inline README.md:108]: You need the ending colon after *:note*

File: README.md [hunk 12]
```diff
@@ -307,36 +324,36 @@ Sarah now has her original 10 ADA plus an additional 5 AGIX.
...
-1. Charlie and Mike successfully build their transactions since the UTxOs still exist.
+1. Charlie and Mike both build their transactions since the UTxOs still exist.
 2. Charlie and Mike submit their transaction at the same time.
 3. Charlie's is added to a block first.
 4. When Mike's transaction is then picked to go into a block, the UTxOs no longer exist. The transaction fails without needing to run the swap contracts.
 
-Since Mike's transaction will fail without needing to run the swap script, Mike's collateral is safe. Further, the more available swaps their are, the less likely these "collisions" are.
+Since Mike's transaction will fail without needing to run the swap script, Mike's collateral is safe. Further, the more available swaps there are, the less likely these "collisions" will occur.
+
+:note future iterations of Ouroboros (namely Leios) may allow arbitragers to further limit these collisions by segmenting transactions among sharded mempools.
```
Comment (fallen-icarus) [inline README.md:344]: Need the ending colon for *:note*

File: README.md [hunk 13]
```diff
@@ -372,37 +389,37 @@ For updating open swaps, I was successfully able to update 5 positions and recre
...
 ---
-## Frontend Agnostic
-By using the beacon tokens, it is trivial for any frontend to integrate with Cardano-Swaps. For example, the new Lace wallet by IOHK can integrate the DEX by simply adding support for querying the beacon tokens. They can also add their own user friendly way to create and use swaps. The only requirement is that all frontends agree to use the same beacon token standard. There is no need for risky extensions or dedicated frontends in order to use this DEX.
+## Frontend Agnosticism
+By using the beacon tokens, it is trivial for any frontend to integrate with Cardano-Swaps. For example, the new Lace wallet by IOHK can integrate the DEX by simply adding support for querying the beacon tokens. They can also add their own user friendly way to create and use swaps. The only requirement is that all frontends/users agree to use the same beacon token standard. There is no need for risky extensions or dedicated frontends in order to use this DEX.
 
 ---
 ## Conclusion
-This DEX protocol has all of the desired properties of a DEX:
+The cardano-swaps protocol has all of the desired properties of a highly scalable DEX:
```
Comment (fallen-icarus) [inline README.md:408]: Need to mention delegation here too.

Comment (fallen-icarus) [reply README.md:408]: Also since you dropped the list, you should replace the ending colon with a period.

File: README.md [hunk 14]
```diff
@@ -372,37 +389,37 @@ For updating open swaps, I was successfully able to update 5 positions and recre
...
 
 ---
 ## FAQ
 
-### If all users have direct access to the spending script for that trading pair, why are reference scripts required?
+#### If all users have direct access to the spending script for that trading pair, why are reference scripts used?
+
+To disincentivize beacon bloat. 
+
+The reason is to incentivize users to actually use the DEX as intended. If there were no deposits, users can create a live address with a beacon token and then just leave it open even if they aren't actually using the address anymore (like if there are no swappable assets in it). This "zombie" address will still appear when querying the beacon tokens even though other users can't actually do anything with it. By requiring the deposits, users are incentivized to close unused addresses in order to get the deposits back.
```
Comment (fallen-icarus) [inline README.md:420]: I could technically use the beacon without reference scripts and just force that the beacon is minted to a utxo containing at least 20 ADA. So this doesn't really answer "why reference scripts.." I forgot to explain it in the original. The idea was:
1) Force a deposit - you can't add a reference script without a minimum amount of ada.
2) Have reference scripts available for use - using reference scripts is usually cheaper than using a local script (although it generally requires more tx memory)

Number 2 may not be necessary since `cardano-reference-scripts` allows sharing reference scripts.

**Comment by fallen-icarus** [comment]
I really think you need to come up with a different "abstract". You are solely focuses right now on scalability and concurrency, but the real motivation was delegation control. The path was: 1) make it decentralized 2) make it scalable. It just so happens that the same design that allows for full delegation control is the same design that allows for organic scalability.

> cardano-swaps is a proof-of-concept implementation of massively scalable and concurrent DEX protocol for the Cardano Settlement Layer (CSL). It solves many of the pitfalls of current DEX implementations by empowering users to deploy their own (and interact with each others') script addresses. By doing so, users maintain fully custody of their assets at all times, and can elect if/when to upgrade their addresses to new contract standards.

This abstract makes it sound like the path of discovery was reverse. Perhaps something like this is a better abstract:

cardano-swaps is a proof-of-concept implementation of a DEX that is organically scalable and concurrent without needing to use batchers, and it doesn't require users to give up custody or delegation control of their assets while using it. It just so happens that, thanks to the eUTxO model on Cardano, organic scalability and full delegation control go hand-in-hand.

**Comment by zhekson1** [comment]
full delegation control goes hand in hand with "users maintain full custody of their assets at all times". For the abstract, summarizing features/overview is more important than path of discovery.

**Comment by fallen-icarus** [comment]
> full delegation control goes hand in hand with "users maintain full custody of their assets at all times".

This is not true. Custody deals with payment keys/scripts while delegation control deals with staking keys/scripts. You can maintain custody without controlling delegation and rewards.

**Comment by fallen-icarus** [comment]
I think it is fine to focus on scalability without needing batchers but then 1) you should really emphasize delegation control as a further bonus/incentive to use this design and 2) you need smooth transitions for each section. Some of them speak as if delegation control is the main thing.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhekson1) [reply README.md:45]: to your second point - no matter how decentralized or how many batchers there are or how well they work, cardano-swaps IS different in that it maximizes eUTXO concurrency. Other DEXes use batchers because addresses are centralized-ish. cardano-swaps does not. So the former is definitely less scalable than the latter.

File: README.md (hunk 2 above)
Comment (zhekson1) [reply README.md:53]: Disagree. LPs were a necessity for EVM paradigm - no other way. For Cardano however, centralized addresses are used because of the routing problem, which is solved by Beacon Tokens. LPs follow from this limitation. 

Composable p2p swaps --> no LPs necessary, not the other way around

File: README.md (hunk 5 above)
Comment (zhekson1) [reply README.md:63]: The redundant execution is a fixable problem. 

Again, I disagree that batchers are just as scalable. They may be for a time, but their resources do not grow proportionally with # of users, whereas cardano-swaps is only limited by off chain querying, and not batcher resources

File: README.md (hunk 8 above)
Comment (zhekson1) [reply README.md:94]: like at all ever for any token?

File: README.md (hunk 14 above)
Comment (zhekson1) [reply README.md:420]: how is forcing a deposit enforced? is it possible to alter the code, not be forced to deposit, and still participate in the protocol?

File: README.md (hunk 8 above)
Comment (fallen-icarus) [reply README.md:94]: At all. The minting policy only allows minting one kind of token: the token with the Policy ID for the minting script and the empty token name. Tokens with any other token name are not mintable. Further, the newly minted token *must* go to cardano-swaps address with a valid staking credential. If the address doesn't have one, the minting policy is guaranteed to fail. Therefore nothing can be minted.

In short, either the proper beacon token is minted or nothing is minted.

File: README.md (hunk 14 above)
Comment (fallen-icarus) [reply README.md:420]: The beacon policy can see the entire tx. It looks through the outputs to find what utxo it is going to be stored in. With this information it checks:
1) The utxo is in the proper address.
2) The utxo has the proper reference script for the Dapp.

As a consequence of number 2, since the reference script takes up space, the ledger forces it to be stored with a minimum amount of ADA. This is the deposit. I could change this so that, instead of requiring the utxo to have the proper reference script, the utxo must have a minimum of 20 ADA. The end result is the same.

> is it possible to alter the code, not be forced to deposit, and still participate in the protocol?
Yes, if it doesn't require number 2 and doesn't require the utxo to have a minimum value, then there is no required deposit.

Comment (zhekson1) [reply README.md:420]: ok, then wouldn't it be trivial for an attacker to create a large number of DDOS/zombie addresses, disrupting off-chain querying? 

If so, is there a way to modify/standardize the beacon policy such there is such a thing as "valid" beacons and that they can only be minted to addresses with a certain DDOS-preventative deposit?

Comment (fallen-icarus) [reply README.md:420]: > wouldn't it be trivial for an attacker to create a large number of DDOS/zombie addresses, disrupting off-chain querying?

Only if I removed the deposit check. Currently, it is impossible to mint a beacon to a dApp address that does not also deposit about 25 ADA (from the size of the reference script). That means if someone were to create 1,000 zombie addresses, they would need to lock up 25,000 ADA for the lifespan of the attack. The dApp script ensures that the deposit can only be reclaimed if the beacon is burned which would stop the attack.

> is there a way to modify/standardize the beacon policy so that there is such a thing as "valid" beacons and that they can only be minted to addresses with an agreed-upon DDOS-preventative deposit?

If the beacon exists at all, then the deposit was guaranteed to be made and the beacon is guaranteed to be protected by the proper spending script.

For generalizing the beacons, we should probably mention in the beacon CIP the need for a deposit to mitigate this attack.

Comment (zhekson1) [reply README.md:420]: I'm saying what if someone altered the code, removed the deposit check, and minted beacons without the deposit?

Are you saying that such beacons would be different from (and always incompatible with) beacons minted by those using scripts *with* deposits?

Comment (fallen-icarus) [reply README.md:420]: > what if someone altered the code, removed the deposit check, and minted beacons without the deposit?

Different code -> different minting policy -> different beacon policy ID -> entirely different beacon

> Are you saying that such beacons would be different from (and always incompatible with) beacons minted by those using scripts with deposits?

Yes. The beacon policy ID is the hash of the minting policy. If the code is different, then the minting policy ID is guaranteed to be different. So even if they use the same token name, the policy ID would be different and therefore it would be an entirely different token.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhekson1) [reply README.md:45]: will elaborate on that in the discussion instead of motivation

**Comment by zhekson1** [comment]
Reorganized for better flow, have not yet edited much in Specification section. Everything else is more or less coming together. Will continue to touch up here and there.

File: README.md [hunk 15]
```diff
@@ -11,40 +11,30 @@ The Getting Started instructions can be found [here](GettingStarted.md).
...
+- [Benchmarks and Fee Estimations](<#benchmarks and fee estimations (ymmv)>)
 - [FAQ](#faq)
-  - [Why Are Reference Scripts Required?](#if-all-users-have-direct-access-to-the-spending-script-for-that-trading-pair-why-are-reference-scripts-required)
+- [Conclusion](#conclusion)
 
 ---
 ## Abstract
-`cardano-swaps` is a proof-of-concept implementation of massively scalable and concurrent DEX protocol for the Cardano Settlement Layer (CSL). It solves many of the pitfalls of current DEX implementations by empowering users to deploy their own (and interact with each others') script addresses. By doing so, users maintain fully custody of their assets at all times, and can elect if/when to upgrade their addresses to new contract standards.
+`cardano-swaps` is a proof-of-concept implementation of an *organically* scalable DEX protocol for the Cardano Settlement Layer (CSL). It solves many of the pitfalls of current DEX implementations by empowering users to deploy their own (and interact with each others') script addresses. By doing so, users always maintain spending *and* delegation custody of their assets, and can elect if/when to upgrade their addresses to new contract standards.
```
Comment (fallen-icarus) [inline README.md:32]: "spending *and* delegation custody" sounds awkward here. People don't usually say "spending custody" or "delegation custody". I think "custody" should be replaced with "control".

File: README.md [hunk 16]
```diff
@@ -58,60 +48,55 @@ One consequence of centralized script addresses is the necessity for liquidity p
...
+Cardano-Swaps achieves batcher/router-free scalability *with* delegation control via a novel combination of user-controlled script addresses and Beacon Tokens (both are expanded upon below).
 
 ### Programmable Swaps
-Programmable Swaps are a better design choice than liquidity pools. They were first proposed by Axo (formerly known as Maladex). However, in Axo's original [whitepaper](https://www.axo.trade/whitepaper.pdf), there is no mention of giving users *full* delegation control of their own assets. 
+Programmable Swaps are a more promising design choice than liquidity pools. They were first proposed by Axo in their original [whitepaper](https://www.axo.trade/whitepaper.pdf), although there is no mention of giving users *full* delegation control over their assets. In essence, programmable swaps are simply locked UTxOs guarded by a user's own script address. They may be used as inputs, if and only if the resulting TX outputs adhere to the swap logic of the script address. This design pattern scales naturally, since there must be at *least* as many swap addresses as there are users.
```
Comment (fallen-icarus) [inline README.md:60]: > In essence, ...

Are you talking about Axo's version or cardano-swaps? I don't think Axo gives each user their own script address.

**Comment by zhekson1** [comment]
Ready for merge if you are. Decent first draft.

File: README.md [hunk 17]
```diff
@@ -5,111 +5,105 @@
...
+  - [The DEX's Inline Datum](<#Specification/#the dex's inline datum>)
+  - [Swap Contract Logic](<#Specification/#swap contract logic>)
+    - [Close Redeemer](<#Specification/##swap contract logic/###close redeemer>)
+    - [Update Redeemer](<#Specification/#swap contract logic/#update redeemer>)
+    - [Swap Redeemer](<#specification/#swap contract logic/#swap redeemer>)
+- [Features Discussion](<#Features Discussion>)
+- [Benchmarks and Fee Estimations](<#benchmarks and fee estimations (ymmv)>)
 - [FAQ](#faq)
-  - [Why Are Reference Scripts Required?](#if-all-users-have-direct-access-to-the-spending-script-for-that-trading-pair-why-are-reference-scripts-required)
+- [Conclusion](#conclusion)
```
Comment (fallen-icarus) [inline README.md:26]: A lot of the table of content links are broken.

File: README.md [hunk 18]
```diff
@@ -5,111 +5,105 @@
...
 
-  1. Composable atomic swaps.
-  2. Users maintain custody of their assets at all times.
-  3. Naturally concurrent and gets more concurrent the more users there are. No batchers are required.
-  4. Liquidity naturally spreads to all trading pairs instead of being siloed into specific trading pairs.
-  5. There is no impermanent loss.
-  6. ADA is all you need to interact with the DEX.
-  7. Upgradability can happen democratically.
-  8. Easy to integrate into any frontend.
+:note: the novel use of *Beacon Tokens* for "tagging" on-chain data can be generalized for many dApps, not just DEXes. More on this in the Beacon Token CIP (**LINK HERE**)
```
Comment (fallen-icarus) [inline README.md:65]: Need to add beacon cip link. You can just link to the PR for now.

File: README.md [hunk 19]
```diff
@@ -5,111 +5,105 @@
...
+### Minting and Using Beacon Tokens
+Beacon Tokens are used to tag `cardano-swaps` addresses so they are readily queryable via an off-chain API, such as Koios or Blockfrost. It is relatively straightforward to find all addresses that contain a specific native token. Here are some examples:
 
 | Task | Koios Api | Blockfrost Api |
 |--|--|--|
 | Addresses with a beacon | [api](https://api.koios.rest/#get-/asset_address_list) | [api](https://docs.blockfrost.io/#tag/Cardano-Assets/paths/~1assets~1%7Basset%7D~1addresses/get)|
 | UTxOs at the address | [api](https://api.koios.rest/#post-/address_info) | [api](https://docs.blockfrost.io/#tag/Cardano-Addresses/paths/~1addresses~1%7Baddress%7D~1utxos/get)|
 
-The utxos api also returns which utxos contain reference scripts. This is how users can remotely execute swap contracts.
+The UTxO query also returns which UTxOs contain reference scripts, allowing users execute their swaps.
```
Comment (fallen-icarus) [inline README.md:101]: allowing users *to remotely* execute their swaps

File: README.md [hunk 20]
```diff
@@ -245,25 +245,29 @@ The `Swap` redeemer checks all of the assets leaving the swap address and all of
...
 What about converting 10 ADA, 5 DUST, and 3 WMT into 16 AGIX and 1 of your favorite NFTs? Piece of cake!
 
-By composing these atomic swaps, in one transaction, you can easily and securely do:
-``` Haskell
-many assets -> many assets
-```
+By composing these swaps in one transaction, many-to-many multi-asset swaps are possible. The only limits are the maximum transaction limits for Cardano.
 
-The only limits are the maximum transaction limits for Cardano.
+**(Maybe: references to Plutus redundant execution CPS)**
```
Comment (fallen-icarus) [inline README.md:265]: I don't think this is necessary since it is mentioned later.

File: README.md [hunk 21]
```diff
@@ -372,37 +370,54 @@ For updating open swaps, I was successfully able to update 5 positions and recre
...
-7. Any frontend can easily add Cardano-Swap support.
+- If the staking credential is a pubkey, then the staking pubkey must sign the transaction.
+- If the staking credential is a script, then the script must be executed in the same transaction.
+- the staking credential effectively *becomes* the "owner" for all actions except for the actual swap execution, in which case the spending credential is used directly.
 
-Thanks to using beacon tokens, decentralization is no longer limited by the design of DEXs. Instead, the limiting factor is now the off-chain querying. However, innovations in this space are still in the early days. The Koios api is an example of a more decentralized off-chain platform. As the technology improves, the decentralization of this DEX will only improve.
+ :note: It is possible to execute a staking script even if 0 ADA is withdrawn from a reward address. The only requirement to use staking scripts like this is that the associated stake address must be registered and delegated. Stake addresses can be utilized this way as soon as the registration+delegation transaction is added to the chain. There is no epoch waiting-period.
```
Comment (fallen-icarus) [inline README.md:390]: The :note: aren't being converted. I think it needs to either be memo or notebook ([emoji list](https://github.com/markdown-templates/markdown-emojis)).

File: README.md [hunk 22]
```diff
@@ -372,37 +370,54 @@ For updating open swaps, I was successfully able to update 5 positions and recre
...
 
----
-## FAQ
+#### Why not just give each user their own spending script?
+`cardano-swaps` relies on the usage of [beacon tokens](<#specification/##Beacon Tokens>) to "tag" addresses, which demarcates them as *distinct* for efficient off-chain querying/aggregation. However, **address distinction is impossible if each script address is composed of completely unique credentials.** In other words, for Beacon Tokens to work properly, there must be a clear distinction between addresses participating in a given dApp and all other addresses. `cardano-swaps` leverages the dual payment-staking credentials of Cardano addresses to maintain address-distinction (shared spending scripts) without sacrificing self-custody (unique staking keys/scripts). For further clarification, refer to the [Beacon Token](<#specification/##Beacon Tokens>) section below, or to the Beacon Token CIP (**link here**)
```
Comment (fallen-icarus) [inline README.md:393]: Just link to the beacon token PR for now.

**Comment by fallen-icarus** [comment]
I think it can be merged once the above minor edits are done.

File: README.md (hunk 19 above)
Comment (zhekson1) [reply README.md:101]: wdym remotely? not clear

Comment (fallen-icarus) [reply README.md:101]: My thought is that the swapper is referencing the script from a far but I guess it isn't necessary. Just add the *to* for grammar.
